Public/Src/Cache/ContentStore/Distributed/NuCache/ContentLocationDatabase.cs

```diff
@@ -372,68 +372,87 @@ private BoolResult GarbageCollectContent(OperationContext context)
-                    if (context.Token.IsCancellationRequested)
+                    // Use double-checked locking to usually avoid locking, but still
+                    // be safe in case we are in a race to update content location data.
+                    lock (GetLock(hash))
                     {
-                        break;
-                    }
+                        if (!TryGetEntryCore(context, hash, out entry))
+                        {
+                            continue;
+                        }
+                        filteredEntry = FilterInactiveMachines(entry);
 
-                    if (!TryGetEntryCore(context, hash, out var entry))
-                    {
-                        continue;
+                        if (filteredEntry.Locations.Count == 0)
+                        {
+                            // If there are no good locations, remove the entry.
+                            removedEntries++;
+                            Counters[ContentLocationDatabaseCounters.TotalNumberOfCollectedEntries].Increment();
+                            Delete(context, hash);
+                            LogEntryDeletion(context, hash, entry, OperationReason.GarbageCollect, replicaCount);
+                        }
+                        else if(filteredEntry.Locations.Count != entry.Locations.Count)
+                        {
+                            // If there are some bad locations, remove them.
+                            Counters[ContentLocationDatabaseCounters.TotalNumberOfCleanedEntries].Increment();
+                            Store(context, hash, filteredEntry);
+                            _nagleOperationTracer.Enqueue((hash, EntryOperation.RemoveMachine, OperationReason.GarbageCollect, entry.Locations.Count - filteredEntry.Locations.Count));
+                        }
                     }
+                }
 
-                    var replicaCount = entry.Locations.Count;
-
-                    uniqueContentCount++;
-                    uniqueContentSize += entry.ContentSize;
-                    totalContentSize += entry.ContentSize * replicaCount;
-                    totalContentCount += replicaCount;
-
-                    var filteredEntry = FilterInactiveMachines(entry);
-                    if (filteredEntry.Locations.Count == 0)
-                    {
-                        removedEntries++;
-                        Counters[ContentLocationDatabaseCounters.TotalNumberOfCollectedEntries].Increment();
-                        Delete(context, hash);
-                        LogEntryDeletion(context, hash, entry, OperationReason.GarbageCollect, replicaCount);
-                    }
-                    else if (filteredEntry.Locations.Count != entry.Locations.Count)
-                    {
-                        Counters[ContentLocationDatabaseCounters.TotalNumberOfCleanedEntries].Increment();
-                        Store(context, hash, entry);
+                totalEntries++;

```

Comment:
>totalEntries [](start = 16, length = 12)

This is a functional change to the semantics of `totalEntries`. Before your change, it counted the number of values in `EnumerateSortedKeys(context)`. After your change, it effectively counts the number of values such that `TryGetEntryCore` returns true.